### ISSUE TITLE
Unable to connect bluetooth headset after restarting bluez or rebooting

### ISSUE THREAD

**Issue by bobvanderlinden** [bluez/bluez#973]
I'm quite confused by this problem, as I thought this was a regression, but I've also tried reverting my system an older snapshot, but the same problem occurs there too now.

Basically, after a reboot the headset cannot connect to the laptop and the other way around.
The exact same behaviour happens without rebooting by restarting `bluetooth.service`.

To reproduce the issue I've done the following steps:

1 `systemctl stop bluetooth.service`
2. (Attempt to) Remove all state from bluez: `sudo rm -rf /var/lib/bluetooth/*`
3. `systemctl start bluetooth.service`
4. Start `blueberry`
5. Bring my headset into pairing mode.
6. Connect to headset in `blueberry`
7. Notice that headset works fine. Also disconnecting and connecting from the laptop and from the headset works fine.
8. Turn off the headset
9. `systemctl restart bluetooth.service`
10. (for the logs below) Start `btmon`
11. Turn on the headset.
12. Notice the headset fails to connect to the laptop. The headset removes the laptop from its paired devices.
13. Stop `btmon`

<details>
<summary>btmon log</summary>

```
Bluetooth monitor ver 5.76
btmon[16782]: = Note: Linux version 6.6.54 (x86_64)                                              0.671997
btmon[16782]: = Note: Bluetooth subsystem version 2.22                                           0.672000
= New Index: 90:65:84:C6:1B:40 (Primary,USB,hci0)                                         [hci0] 0.672001
= Open Index: 90:65:84:C6:1B:40                                                           [hci0] 0.672002
= Index Info: 90:65:84:C6:1B:40 (Intel Corp.)                                             [hci0] 0.672002
bluetoothd[16591]: @ MGMT Open: bluetoothd (privileged) version 1.22                    {0x0001} 0.672004
> HCI Event: Connect Request (0x04) plen 10                                            #1 [hci0] 9.744671
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Class: 0x240418
          Major class: Audio/Video (headset, speaker, stereo, video, vcr)
          Minor class: Headphones
          Rendering (Printing, Speaker)
          Audio (Speaker, Microphone, Headset)
        Link type: ACL (0x01)
< HCI Command: Accept Connection Request (0x01|0x0009) plen 7                          #2 [hci0] 9.744897
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Role: Central (0x00)
> HCI Event: Command Status (0x0f) plen 4                                              #3 [hci0] 9.745525
      Accept Connection Request (0x01|0x0009) ncmd 1
        Status: Success (0x00)
> HCI Event: Role Change (0x12) plen 8                                                 #4 [hci0] 9.923506
        Status: Success (0x00)
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Role: Central (0x00)
> HCI Event: Vendor (0xff) plen 4                                                     #5 [hci0] 10.109708
      Intel PTT Switch Notification (0x26)
        Handle: 256
        Packet type table: Enhanced data rate (0x01)
> HCI Event: Connect Complete (0x03) plen 11                                          #6 [hci0] 10.112500
        Status: Success (0x00)
        Handle: 256
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Link type: ACL (0x01)
        Encryption: Disabled (0x00)
< HCI Command: Read Remote Supported Features (0x01|0x001b) plen 2                    #7 [hci0] 10.112815
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
btmon[16782]: @ RAW Open: btmon (privileged) version 2.22                              {0x0002} 10.113232
btmon[16782]: @ RAW Close: btmon                                                       {0x0002} 10.113243
> HCI Event: Command Status (0x0f) plen 4                                             #8 [hci0] 10.113454
      Read Remote Supported Features (0x01|0x001b) ncmd 1
        Status: Success (0x00)
< HCI Command: Write Scan Enable (0x03|0x001a) plen 1                                 #9 [hci0] 10.113560
        Scan enable: No Scans (0x00)
> HCI Event: Command Complete (0x0e) plen 4                                          #10 [hci0] 10.114501
      Write Scan Enable (0x03|0x001a) ncmd 2
        Status: Success (0x00)
> ACL Data RX: Handle 256 flags 0x02 dlen 12                                         #11 [hci0] 10.121500
      L2CAP: Connection Request (0x02) ident 1 len 4
        PSM: 1 (0x0001)
        Source CID: 128
> HCI Event: Max Slots Change (0x1b) plen 3                                          #12 [hci0] 10.129429
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Max slots: 5
> HCI Event: Read Remote Supported Features (0x0b) plen 11                           #13 [hci0] 10.131428
        Status: Success (0x00)
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Features: 0xff 0xff 0x8f 0xfe 0xdb 0xff 0x5b 0x87
          3 slot packets
          5 slot packets
          Encryption
          Slot offset
          Timing accuracy
          Role switch
          Hold mode
          Sniff mode
          Park state
          Power control requests
          Channel quality driven data rate (CQDDR)
          SCO link
          HV2 packets
          HV3 packets
          u-law log synchronous data
          A-law log synchronous data
          CVSD synchronous data
          Paging parameter negotiation
          Power control
          Transparent synchronous data
          Broadcast Encryption
          Enhanced Data Rate ACL 2 Mbps mode
          Enhanced Data Rate ACL 3 Mbps mode
          Enhanced inquiry scan
          Interlaced inquiry scan
          Interlaced page scan
          RSSI with inquiry results
          Extended SCO link (EV3 packets)
          EV4 packets
          EV5 packets
          AFH capable peripheral
          AFH classification peripheral
          LE Supported (Controller)
          3-slot Enhanced Data Rate ACL packets
          5-slot Enhanced Data Rate ACL packets
          Sniff subrating
          Pause encryption
          AFH capable central
          AFH classification central
          Enhanced Data Rate eSCO 2 Mbps mode
          Enhanced Data Rate eSCO 3 Mbps mode
          3-slot Enhanced Data Rate eSCO packets
          Extended Inquiry Response
          Simultaneous LE and BR/EDR (Controller)
          Secure Simple Pairing
          Encapsulated PDU
          Non-flushable Packet Boundary Flag
          Link Supervision Timeout Changed Event
          Inquiry TX Power Level
          Enhanced Power Control
          Extended features
< HCI Command: Read Remote Extended Features (0x01|0x001c) plen 3                    #14 [hci0] 10.131572
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Page: 1
> HCI Event: Command Status (0x0f) plen 4                                            #15 [hci0] 10.132422
      Read Remote Extended Features (0x01|0x001c) ncmd 1
        Status: Success (0x00)
> HCI Event: Read Remote Extended Features (0x23) plen 13                            #16 [hci0] 10.136429
        Status: Success (0x00)
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Page: 1/2
        Features: 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00
          Secure Simple Pairing (Host Support)
          LE Supported (Host)
< HCI Command: Remote Name Request (0x01|0x0019) plen 10                             #17 [hci0] 10.136565
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Page scan repetition mode: R2 (0x02)
        Page scan mode: Mandatory (0x00)
        Clock offset: 0x0000
< ACL Data TX: Handle 256 flags 0x00 dlen 10                                         #18 [hci0] 10.136586
      L2CAP: Information Request (0x0a) ident 1 len 2
        Type: Extended features supported (0x0002)
@ MGMT Event: Device Connected (0x000b) plen 18                                 {0x0001} [hci0] 10.136596
        BR/EDR Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Flags: 0x00000000
        Data length: 5
        Class: 0x240418
          Major class: Audio/Video (headset, speaker, stereo, video, vcr)
          Minor class: Headphones
          Rendering (Printing, Speaker)
          Audio (Speaker, Microphone, Headset)
< ACL Data TX: Handle 256 flags 0x00 dlen 16                                         #19 [hci0] 10.136619
      L2CAP: Connection Response (0x03) ident 1 len 8
        Destination CID: 64
        Source CID: 128
        Result: Connection pending (0x0001)
        Status: No further information available (0x0000)
< ACL Data TX: Handle 256 flags 0x00 dlen 10                                         #20 [hci0] 10.136628
      L2CAP: Information Request (0x0a) ident 2 len 2
        Type: Extended features supported (0x0002)
> HCI Event: Command Status (0x0f) plen 4                                            #21 [hci0] 10.137421
      Remote Name Request (0x01|0x0019) ncmd 1
        Status: Success (0x00)
> HCI Event: Number of Completed Packets (0x13) plen 5                               #22 [hci0] 10.150419
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #18: len 10 (6 Kb/s)
        Latency: 13 msec (13-13 msec ~13 msec)
> HCI Event: Number of Completed Packets (0x13) plen 5                               #23 [hci0] 10.151431
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #19: len 16 (9 Kb/s)
        Latency: 14 msec (13-14 msec ~14 msec)
> ACL Data RX: Handle 256 flags 0x02 dlen 16                                         #24 [hci0] 10.154887
      L2CAP: Information Response (0x0b) ident 1 len 8
        Type: Extended features supported (0x0002)
        Result: Success (0x0000)
        Features: 0x000000b8
          Enhanced Retransmission Mode
          Streaming Mode
          FCS Option
          Fixed Channels
> HCI Event: Remote Name Req Complete (0x07) plen 255                                #25 [hci0] 10.159419
        Status: Success (0x00)
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Name: Hipsterapparaat
> ACL Data RX: Handle 256 flags 0x02 dlen 16                                         #26 [hci0] 10.159873
      L2CAP: Information Response (0x0b) ident 2 len 8
        Type: Extended features supported (0x0002)
        Result: Success (0x0000)
        Features: 0x000000b8
          Enhanced Retransmission Mode
          Streaming Mode
          FCS Option
          Fixed Channels
< ACL Data TX: Handle 256 flags 0x00 dlen 10                                         #27 [hci0] 10.159951
      L2CAP: Information Request (0x0a) ident 3 len 2
        Type: Fixed channels supported (0x0003)
> HCI Event: Number of Completed Packets (0x13) plen 5                               #28 [hci0] 10.160423
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #20: len 10 (3 Kb/s)
        Latency: 23 msec (13-23 msec ~19 msec)
> HCI Event: Number of Completed Packets (0x13) plen 5                               #29 [hci0] 10.163420
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #27: len 10 (26 Kb/s)
        Latency: 3 msec (3-23 msec ~11 msec)
> ACL Data RX: Handle 256 flags 0x02 dlen 20                                         #30 [hci0] 10.173965
      L2CAP: Information Response (0x0b) ident 3 len 12
        Type: Fixed channels supported (0x0003)
        Result: Success (0x0000)
        Channels: 0x0000000000000006
          L2CAP Signaling (BR/EDR)
          Connectionless reception
< ACL Data TX: Handle 256 flags 0x00 dlen 16                                         #31 [hci0] 10.174046
      L2CAP: Connection Response (0x03) ident 1 len 8
        Destination CID: 64
        Source CID: 128
        Result: Connection successful (0x0000)
        Status: No further information available (0x0000)
< ACL Data TX: Handle 256 flags 0x00 dlen 23                                         #32 [hci0] 10.174061
      L2CAP: Configure Request (0x04) ident 4 len 15
        Destination CID: 128
        Flags: 0x0000
        Option: Retransmission and Flow Control (0x04) [mandatory]
          Mode: Basic (0x00)
          TX window size: 0
          Max transmit: 0
          Retransmission timeout: 0
          Monitor timeout: 0
          Maximum PDU size: 0
> HCI Event: Number of Completed Packets (0x13) plen 5                               #33 [hci0] 10.178507
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #31: len 16 (32 Kb/s)
        Latency: 4 msec (3-23 msec ~7 msec)
> HCI Event: Number of Completed Packets (0x13) plen 5                               #34 [hci0] 10.185503
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #32: len 23 (16 Kb/s)
        Latency: 11 msec (3-23 msec ~9 msec)
> ACL Data RX: Handle 256 flags 0x02 dlen 16                                         #35 [hci0] 10.186974
      L2CAP: Configure Request (0x04) ident 2 len 8
        Destination CID: 64
        Flags: 0x0000
        Option: Maximum Transmission Unit (0x01) [mandatory]
          MTU: 48
< ACL Data TX: Handle 256 flags 0x00 dlen 18                                         #36 [hci0] 10.187132
      L2CAP: Configure Response (0x05) ident 2 len 10
        Source CID: 128
        Flags: 0x0000
        Result: Success (0x0000)
        Option: Maximum Transmission Unit (0x01) [mandatory]
          MTU: 48
> HCI Event: Number of Completed Packets (0x13) plen 5                               #37 [hci0] 10.191416
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #36: len 18 (36 Kb/s)
        Latency: 4 msec (3-23 msec ~6 msec)
> ACL Data RX: Handle 256 flags 0x02 dlen 14                                         #38 [hci0] 10.193968
      L2CAP: Configure Response (0x05) ident 4 len 6
        Source CID: 64
        Flags: 0x0000
        Result: Success (0x0000)
> ACL Data RX: Handle 256 flags 0x02 dlen 22                                         #39 [hci0] 10.201990
      Channel: 64 len 18 [PSM 1 mode Basic (0x00)] {chan 0}
      SDP: Service Search Attribute Request (0x06) tid 1 len 13
        Search pattern: [len 5]
          Sequence (6) with 3 bytes [8 extra bits] len 5
            UUID (3) with 2 bytes [0 extra bits] len 3
              Advanced Audio Distribution (0x110d)
        Max record count: 32
        Attribute list: [len 5]
          Sequence (6) with 3 bytes [8 extra bits] len 5
            Unsigned Integer (1) with 2 bytes [0 extra bits] len 3
              0x0009
        Continuation state: 0
bluetoothd[16591]: < ACL Data TX: Handle 256 flags 0x00 dlen 44                      #40 [hci0] 10.202642
      Channel: 128 len 40 [PSM 1 mode Basic (0x00)] {chan 0}
      SDP: Service Search Attribute Response (0x07) tid 1 len 35
        Attribute bytes: 32
          Attribute list: [len 13] {position 0}
            Attribute: Bluetooth Profile Descriptor List (0x0009) [len 2]
              Sequence (6) with 6 bytes [8 extra bits] len 8
                UUID (3) with 2 bytes [0 extra bits] len 3
                  Advanced Audio Distribution (0x110d)
                Unsigned Integer (1) with 2 bytes [0 extra bits] len 3
                  0x0103
          Attribute list: [len 13] {position 1}
            Attribute: Bluetooth Profile Descriptor List (0x0009) [len 2]
              Sequence (6) with 6 bytes [8 extra bits] len 8
                UUID (3) with 2 bytes [0 extra bits] len 3
                  Advanced Audio Distribution (0x110d)
                Unsigned Integer (1) with 2 bytes [0 extra bits] len 3
                  0x0103
        Continuation state: 0
> HCI Event: Number of Completed Packets (0x13) plen 5                               #41 [hci0] 10.206499
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #40: len 44 (117 Kb/s)
        Latency: 3 msec (3-23 msec ~5 msec)
        Channel: 128 [PSM 1 mode Basic (0x00)] {chan 0}
        Channel Latency: 3 msec (3-3 msec ~3 msec)
> ACL Data RX: Handle 256 flags 0x02 dlen 12                                         #42 [hci0] 10.210038
      L2CAP: Disconnection Request (0x06) ident 3 len 4
        Destination CID: 64
        Source CID: 128
< ACL Data TX: Handle 256 flags 0x00 dlen 12                                         #43 [hci0] 10.210177
      L2CAP: Disconnection Response (0x07) ident 3 len 4
        Destination CID: 64
        Source CID: 128
> HCI Event: Number of Completed Packets (0x13) plen 5                               #44 [hci0] 10.214432
        Num handles: 1
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Count: 1
        #43: len 12 (24 Kb/s)
        Latency: 4 msec (3-23 msec ~4 msec)
> HCI Event: Link Key Request (0x17) plen 6                                          #45 [hci0] 10.222516
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
< HCI Command: Link Key Request Negative Reply (0x01|0x000c) plen 6                  #46 [hci0] 10.222647
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
> HCI Event: Command Complete (0x0e) plen 10                                         #47 [hci0] 10.223497
      Link Key Request Negative Reply (0x01|0x000c) ncmd 1
        Status: Success (0x00)
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
> HCI Event: Disconnect Complete (0x05) plen 4                                       #48 [hci0] 10.648720
        Status: Success (0x00)
        Handle: 256 Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Reason: Authentication Failure (0x05)
@ MGMT Event: Device Disconnected (0x000c) plen 8                               {0x0001} [hci0] 10.648907
        BR/EDR Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
        Reason: Unspecified (0x00)
< HCI Command: Write Scan Enable (0x03|0x001a) plen 1                                #49 [hci0] 10.669090
        Scan enable: Page Scan (0x02)
> HCI Event: Command Complete (0x0e) plen 4                                          #50 [hci0] 10.670490
      Write Scan Enable (0x03|0x001a) ncmd 2
        Status: Success (0x00)
```

</details>

[btsnoop.log](https://github.com/user-attachments/files/17318435/btsnoop.log)

After this I also tried connecting from laptop to headset:

14. Start `blueberry`
15. Run `journalctl --unit bluetooth --follow`
16. Connect to headset
17. Connecting fails
18. Notice the `Hands-Free Voice gateway failed connect` in the bluetooth logs

<details>
<summary>systemctl --unit bluetooth --follow</summary>

```
Oct 09 23:09:53 nac44250 bluetoothd[22751]: src/profile.c:ext_connect() Hands-Free Voice gateway failed connect to 2C:41:A1:C8:E5:04: Connection refused (111)
Oct 09 23:09:53 nac44250 bluetoothd[22751]: profiles/audio/avdtp.c:avdtp_connect_cb() connect to 2C:41:A1:C8:E5:04: Connection refused (111)
Oct 09 23:09:55 nac44250 bluetoothd[22751]: profiles/audio/avdtp.c:avdtp_connect_cb() connect to 2C:41:A1:C8:E5:04: Connection refused (111)
```

</details>

Instead of `systemctl restart bluetooth`, a reboot will have the same effect(!):

1. Re-pair headset and connect
2. Notice it being able to connect and reconnect
3. Turn off headset
4. Reboot
5. Turn on headset
6. Notice it not being able to connect anymore

Also worth noting: the bluetooth mouse keeps working.

This is on a NixOS system running bluez-5.76.

**Comment by Vudentz** [comment]
```
< HCI Command: Link Key Request Negative Reply (0x01|0x000c) plen 6                  #46 [hci0] 10.222647
        Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
```

Looks like the link-keys cannot be found, is /var/lib/bluetooth being wiped out on service restart? Does that has the same result when running bluetoothctl as pairing agent instead of blueberry?

**Comment by Balssh** [comment]
> ```
> 
> < HCI Command: Link Key Request Negative Reply (0x01|0x000c) plen 6                  #46 [hci0] 10.222647
> 
>         Address: 2C:41:A1:C8:E5:04 (Bose Corporation)
> 
> ```
> 
> 
> 
> Looks like the link-keys cannot be found, is /var/lib/bluetooth being wiped out on service restart? Does that has the same result when running bluetoothctl as pairing agent instead of blueberry?

I'll have to double check tomorrow at my laptop, but I'm positive I'm using bluetoothctl and have same problem. After every second reboot, bluetooth stops working. I also use nixos, but don't think it's related.

**Comment by ProYourName** [comment]
I have a same issue. Pairing using bluetoothctl, blueman and bluedevil had no effect.

**Comment by Vudentz** [comment]
Are you all using nixos? Does it has some special handling over /var/lib/bluetooth? Or perhaps something with systemd .service file, how does it look like? Here are the defaults:

https://github.com/bluez/bluez/blob/master/src/bluetooth.service.in

**Comment by ProYourName** [comment]
> Are you all using nixos?
No, i use arch. Journalctl show many errors like this:
Bluetooth: hci0: ACL packet for unknown connection handle 3804

**Comment by Balssh** [comment]
A little update: got bluez updated to 5.78 and now it seems to be working fine.

**Comment by bobvanderlinden** [comment]
I tried to bisect to older versions of nixos, but it never seemed to work no matter which bluez version. However, I didn't reboot between switching of versions, so it was always with the same kernel. Other NixOS users have also found that switching kernels did fix the issue (https://github.com/NixOS/nixpkgs/issues/346443#issuecomment-2433370903).

It has been resolved for me now.

**Comment by maru-sama** [comment]
Hello, I have the same issue with kernel 6.10.13 and 6.11.5. After an initial pairing I can connect my headset but after a restart of the machine it fails to connect. If I boot up with an older 6.8 kernel I am able to connect immediately without re-pairing. 
My mouse works without any issues with all kernels.
I am currently compiling 6.10.14 to check if it fixes the issue. Attached you will find the btmon log

The machine is running Ubuntu 24.04.
[btmon.log](https://github.com/user-attachments/files/17578888/btmon.log)

**Comment by Vudentz** [comment]
@maru-sama is the link-key present in /var/lib/bluetooth? Does it show as paired in the likes of bluetoothctl?

It would also be great if you guys can collect the btmon logs when bluetoothd is restarted so we can check if this is not due something like:

commit 532f8bcd1c2c4e8112f62e1922fd1703bc0ffce0
Author: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
Date:   Tue Aug 27 14:37:22 2024 -0400

    Revert "Bluetooth: MGMT/SMP: Fix address type when using SMP over BREDR/LE"
    
    This reverts commit 59b047bc98084f8af2c41483e4d68a5adf2fa7f7 which
    breaks compatibility with commands like:
    
    bluetoothd[46328]: @ MGMT Command: Load.. (0x0013) plen 74  {0x0001} [hci0]
            Keys: 2
            BR/EDR Address: C0:DC:DA:A5:E5:47 (Samsung Electronics Co.,Ltd)
            Key type: Authenticated key from P-256 (0x03)
            Central: 0x00
            Encryption size: 16
            Diversifier[2]: 0000
            Randomizer[8]: 0000000000000000
            Key[16]: 6ed96089bd9765be2f2c971b0b95f624
            LE Address: D7:2A:DE:1E:73:A2 (Static)
            Key type: Unauthenticated key from P-256 (0x02)
            Central: 0x00
            Encryption size: 16
            Diversifier[2]: 0000
            Randomizer[8]: 0000000000000000
            Key[16]: 87dd2546ededda380ffcdc0a8faa4597
    @ MGMT Event: Command Status (0x0002) plen 3                {0x0001} [hci0]
          Load Long Term Keys (0x0013)
            Status: Invalid Parameters (0x0d)
    
    Cc: stable@vger.kernel.org
    Link: https://github.com/bluez/bluez/issues/875
    Fixes: 59b047bc9808 ("Bluetooth: MGMT/SMP: Fix address type when using SMP over BREDR/LE")
    Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

**Comment by maru-sama** [comment]
I see the following for the headphone in the info file

```
[LinkKey]
Key=675E8BA2930E28EF88306A0F2D5XXXX <-- modified by me
Type=4
PINLength=0

```
and this is during a restart of bluetooth service

[restart.log](https://github.com/user-attachments/files/17579291/restart.log)

I do not see the errors during the key loading though.

**Comment by maru-sama** [comment]
6.10.14 does not fix the issue I still get the same error
```
 ACL Data TX: Handle 256 flags 0x00 dlen 10                                                                                                                              #24 [hci0] 7.562387
      L2CAP: Information Request (0x0a) ident 2 len 2
        Type: Fixed channels supported (0x0003)
> HCI Event: Number of Completed Packets (0x13) plen 5                                                                                                                    #25 [hci0] 7.568017
        Num handles: 1
        Handle: 256 Address: 04:5D:4B:80:CA:CD (Sony Corporation)
        Count: 1
        #24: len 10 (16 Kb/s)
        Latency: 5 msec (5-10 msec ~8 msec)
> HCI Event: Link Key Request (0x17) plen 6                                                                                                                               #26 [hci0] 7.569074
        Address: 04:5D:4B:80:CA:CD (Sony Corporation)
< HCI Command: Link Key Request Negative Reply (0x01|0x000c) plen 6                                                                                                       #27 [hci0] 7.569275
        Address: 04:5D:4B:80:CA:CD (Sony Corporation)
```

**Comment by maru-sama** [comment]
Booting with the standard 6.8.0 without changing anything works immediately

```
 HCI Event: Link Key Request (0x17) plen 6                                                                                                                              #25 [hci0] 10.426378
        Address: 04:5D:4B:80:CA:CD (Sony Corporation)
< HCI Command: Link Key Request Reply (0x01|0x000b) plen 22                                                                                                              #26 [hci0] 10.426441
        Address: 04:5D:4B:80:CA:CD (Sony Corporation)
        Link key: 675e8ba2930e28ef88306a0f2d5ccafc
> HCI Event: Command Complete (0x0e) plen 10                                                                                                                             #27 [hci0] 10.427354
      Link Key Request Reply (0x01|0x000b) ncmd 1
        Status: Success (0x00)
        Address: 04:5D:4B:80:CA:CD (Sony Corporation)
```

**Comment by maru-sama** [comment]
I can confirm that this fixes the issue. I re-build the ubuntu bluez package with this patch added and the headphone connected immediately after a reboot to 6.11.5. Thank you for your support.

**Comment by maru-sama** [comment]
Does this mean that for newer kernels either bluez 5.78 or a backport of this patch  is required for the BT devices to work correctly, or has this been fixed in the 6.12 release cycle in the kernel?

**Comment by icegood** [comment]
is it the same I have in the latest ubuntu?
https://bugs.launchpad.net/ubuntu/+source/bluez/+bug/2084963

**Comment by Vudentz** [comment]
> Does this mean that for newer kernels either bluez 5.78 or a backport of this patch is required for the BT devices to work correctly, or has this been fixed in the 6.12 release cycle in the kernel?

I think you always gonna need the userspace because there could be invalid keys in the storage and the latest kernel just ignore them which is probably what is happening here, anyway I incline to close this as duplicate of https://github.com/bluez/bluez/issues/875.

**Comment by maru-sama** [comment]
> is it the same I have in the latest ubuntu? https://bugs.launchpad.net/ubuntu/+source/bluez/+bug/2084963

I just updated my notebook and see the same issues. Either updating the package to 5.78 or adding the one patch mentioned earlier should fix this issue.

**Comment by bobvanderlinden** [comment]
Thanks for all the feedback. I think the conclusion is to upgrade or patch. Since this has already been fixed in bluez 5.78 I'll close this issue.
